### PR TITLE
CONSOLE-4269: add custom ConsolePlugin resource details page

### PR DESCRIPTION
**Pull request by rhamilto**
Includes changes from https://github.com/openshift/console/pull/14374, which should merge first.  **Note** https://github.com/openshift/console/pull/14374 includes a bug that needs to be addressed where not Loaded plugins do not appear in the `Console plugins` list.

In collaboration with @vojtechszocs 🥇 

TODOs:

- [ ] improve `Description`, `Version`, `Status`, `Enabled`, and `CSP Violations` popover text | @opayne1, can you help here?

### Demos

Plugin is `Loaded`

https://github.com/user-attachments/assets/08e9b736-12bb-4336-a0bb-72e7cc408657

Plugin is **not** `Loaded` (information from pluginStore is not available)

https://github.com/user-attachments/assets/2d8e1efb-880d-480b-aa61-dc8531ed06cc

### Development setup

1. [Build, start](https://github.com/openshift/console/tree/master/dynamic-demo-plugin#local-development), and [run](https://github.com/openshift/console/blob/4df2ddc4aa941a2916279983fdc27b0a29b64268/frontend/packages/integration-tests-cypress/tests/app/demo-dynamic-plugin.cy.ts#L22) the `console-demo-plugin` locally.
2. Apply the [oc-manifest.yaml](https://github.com/openshift/console/blob/master/dynamic-demo-plugin/oc-manifest.yaml) to the cluster.
3. Visit http://localhost:9000/k8s/cluster/console.openshift.io~v1~ConsolePlugin/console-demo-plugin/

### CODE REVIEW

**Comment by rhamilto** [comment]
@vojtechszocs, I removed the changes to `useTelemetry.ts` from this PR since they are unrelated the scope of the story and opened [a bug](https://issues.redhat.com/browse/OCPBUGS-44511) to address.

File: frontend/packages/console-app/locales/en/console-app.json [hunk 1]
```diff
@@ -178,6 +178,15 @@
   "Namespace where the API configuration secret is located": "Namespace where the API configuration secret is located",
   "Scheduling disabled": "Scheduling disabled",
   "Approval required": "Approval required",
+  "Display Name": "Display Name",
+  "Status": "Status",
+  "The state of the plugin.": "The state of the plugin.",
+  "Enabled": "Enabled",
+  "A boolean to determine whether or not the plugin renders in console.": "A boolean to determine whether or not the plugin renders in console.",
```
Comment (opayne1) [inline frontend/packages/console-app/locales/en/console-app.json:185]: What about shortening it? I think the "whether or not" is implied.

```suggestion
  "A boolean to determine if the plugin renders in console.": "A boolean to determine if the plugin renders in console.",
```

File: frontend/packages/console-app/locales/en/console-app.json [hunk 2]
```diff
@@ -178,6 +178,15 @@
...
+  "CSP Violations": "CSP Violations",
+  "Plugin might have violated the Console Content Security Policy.": "Plugin might have violated the Console Content Security Policy.",
```
Comment (opayne1) [inline frontend/packages/console-app/locales/en/console-app.json:187]: Should this one be similar to the one on line 235? or can we make it a complete sentence/

```suggestion
  "This plugin might have violated the Console Content Security Policy.": "This plugin might have violated the Console Content Security Policy.",
```

File: frontend/packages/console-app/src/components/console-operator/ConsoleOperatorConfig.tsx [hunk 3]
```diff
@@ -103,11 +100,34 @@ const ConsolePluginEnabledStatus: React.FC<ConsolePluginEnabledStatusProps> = ({
...
+    consoleOperatorConfigLoaded,
+    canPatchConsoleOperatorConfig,
+  } = useConsoleOperatorConfigData();
+
   const labels = enabled ? t('console-app~Enabled') : t('console-app~Disabled');
 
   return (
     <>
-      {consoleOperatorConfig && canPatchConsoleOperatorConfig ? (
+      {consoleOperatorConfigLoaded && canPatchConsoleOperatorConfig && !developmentMode ? (
```
Comment (rhamilto) [inline frontend/packages/console-app/src/components/console-operator/ConsoleOperatorConfig.tsx:130]: Fixes regression introduced in https://github.com/openshift/console/pull/14403.  We only want to display the button when not in dev mode.

**Review by opayne1 — COMMENTED** [review]
Made a few tweaks. Let me know your thoughts.

File: frontend/packages/console-app/locales/en/console-app.json (hunk 1 above)
Comment (rhamilto) [reply frontend/packages/console-app/locales/en/console-app.json]: Sorry, @opayne1.  I rewrote this while making some additions and just noticed your comment.  The text now reads, `Whether or not the plugin renders in console.`. WDYT?

File: frontend/packages/console-app/locales/en/console-app.json (hunk 2 above)
Comment (rhamilto) [reply frontend/packages/console-app/locales/en/console-app.json]: Sorry, @opayne1.  I rewrote this while making some additions and just noticed your comment.  The text now reads, `Whether or not the plugin might have violated the Console Content Security Policy.`. WDYT?

**Comment by yanpzhan** [comment]
@rhamilto when edit "Enable" field on plugin details page, the plugin could be enabled repeatedly, we could see multiple plugins with same name in console operator, is it an issue?
Screenshot from 2024-11-14 17-44-35

**Comment by rhamilto** [comment]
> @rhamilto when edit "Enabled" field on plugin details page, the plugin could be enabled repeatedly, we could see multiple plugins with same name in console operator, is it an issue? Screenshot from 2024-11-14 17-44-35

@yanpzhan, I believe that is an existing issue.  It's a bug, but not one that is hugely problematic.  Mind opening a bug?

**Comment by rhamilto** [comment]
> @yanpzhan, I believe that is an existing issue. It's a bug, but not one that is hugely problematic. Mind opening a bug?

This appears to be the result of the patch to the console config taking a while to occur (which happens sporadically).  A partial fix is to disable the `Save` button if the enabled value hasn't been changed in the modal, but that only works reliably if the previous patch has already occurred and the console config is updated.

File: frontend/packages/console-app/console-extensions.json [hunk 4]
```diff
@@ -1891,5 +1891,132 @@
...
+    "properties": {
+      "id": "plugin-csp",
+      "model": {
+        "group": "console.openshift.io",
+        "version": "v1",
+        "kind": "ConsolePlugin"
+      },
+      "column": "right",
+      "title": "%console-app~CSP Violations%",
+      "description": "%console-app~Whether or not the plugin might have violated the Console Content Security Policy.%",
```
Comment (jhadvig) [inline frontend/packages/console-app/console-extensions.json:1985]: nit?:
`...violated the Console's Content Security Policy.` ?

File: frontend/packages/console-app/src/components/console-operator/ConsolePluginBackendDetail.tsx [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import { ServiceModel } from '@console/internal/models';
+import { ConsolePluginKind } from '@console/internal/module/k8s';
+import { DASH } from '@console/shared/src/constants';
+
+const ConsolePluginBackendDetail: React.FC<ConsolePluginBackendDetailProps> = ({
+  obj: {
+    spec: { backend },
+  },
+}) =>
+  // only Service is supported per the ConsolePlugin schema
```
Comment (jhadvig) [inline frontend/packages/console-app/src/components/console-operator/ConsolePluginBackendDetail.tsx:16]: 👍

**Review by jhadvig — APPROVED** [review]
One possible nit, otherwise really nice and clean work 👍 
/lgtm

**Comment by jhadvig** [comment]
QE Approver:
/assign @yapei
Docs Approver:
/assign @opayne1
PX Approver:
/assign @reestr 

Also worth to notice that this PR is cherry-picking [CONSOLE-4264: Notify users of Console plugin related Content Security Policy violations](https://github.com/openshift/console/pull/14374) and also contributing a fix for [issue](https://github.com/openshift/console/pull/14374#issuecomment-2469521566) found by @yapei.

**Comment by yanpzhan** [comment]
Checked on cluster launched against the pr, tested consoleplugin 'Details' page and 'Plugin manifest' tab, features work as expected.
/label qe-approved

**Comment by spadgett** [comment]
Hey, @rhamilto. This is great. I'm happy to see we have a nice details page.

Out of scope for this PR and story, but I had a few thoughts for future improvements:

1. It would be good to show the plugin's supported OCP versions and make it clear when a plugin is disabled since its version range doesn't match the current OCP version. (I believe the supported versions are in the manifest, but would be good to have this on the overview tab. See [plugin manifest properties](https://github.com/openshift/enhancements/blob/master/enhancements/console/dynamic-plugins.md#plugin-manifest-properties).)
2. Should we add a top-level nav items for Console Plugins to the Administration nav section? It would make it more discoverable.
3. Syntax highlighting for the manifest JSON would be nice. Maybe we can use Monaco.
4. I think a table of the extensions used by the plugin would be nice to have in addition to the manifest JSON (but keep the manifest tab for the complete JSON).

@jhadvig @vojtechszocs FYI

**Comment by rhamilto** [comment]
> Hey, @rhamilto. This is great. I'm happy to see we have a nice details page.

Thanks for the feedback, @spadgett.

> 1. It would be good to show the plugin's supported OCP versions and make it clear when a plugin is disabled since its version range doesn't match the current OCP version. (I believe the supported versions are in the manifest, but would be good to have this on the overview tab. See [plugin manifest properties](https://github.com/openshift/enhancements/blob/master/enhancements/console/dynamic-plugins.md#plugin-manifest-properties).)

Easy enough to add.

> 2. Should we add a top-level nav items for Console Plugins to the Administration nav section? It would make it more discoverable.

Agreed. 

> 3. Syntax highlighting for the manifest JSON would be nice. Maybe we can use Monaco.

@vojtechszocs and I started down the path of using Monaco, but decided against it for the initial implementation because the read only mode of the existing YAML editor posed some usability questions that we didn't have time to contend with. 

> 4. I think a table of the extensions used by the plugin would be nice to have in addition to the manifest JSON (but keep the manifest tab for the complete JSON).

Easy enough to add.

**Comment by rhamilto** [comment]
> 1. It would be good to show the plugin's supported OCP versions and make it clear when a plugin is disabled since its version range doesn't match the current OCP version. (I believe the supported versions are in the manifest, but would be good to have this on the overview tab. See [plugin manifest properties](https://github.com/openshift/enhancements/blob/master/enhancements/console/dynamic-plugins.md#plugin-manifest-properties).)

Maybe something like this (note all the details in the video are not accurate such as `Enabled`)?

https://github.com/user-attachments/assets/115f4b54-c895-4cd6-8b63-1875e8ad4d92

**Comment by rhamilto** [comment]
> 3\. Syntax highlighting for the manifest JSON would be nice. Maybe we can use Monaco.

@spadgett, we must've done something wonky with our initial attempt as this appears to be working correctly now.  

I do wonder, is it weird to have two YAML editors?

https://github.com/user-attachments/assets/5b977750-8748-4f5a-8493-03996b4e4fe1

**Comment by rhamilto** [comment]
> 4\. I think a table of the extensions used by the plugin would be nice to have in addition to the manifest JSON (but keep the manifest tab for the complete JSON).

Hmm.  Given the variance amongst the extension properties, this one is a bit tricker to tabularize. 🤔

**Comment by spadgett** [comment]
> > 1. It would be good to show the plugin's supported OCP versions and make it clear when a plugin is disabled since its version range doesn't match the current OCP version. (I believe the supported versions are in the manifest, but would be good to have this on the overview tab. See [plugin manifest properties](https://github.com/openshift/enhancements/blob/master/enhancements/console/dynamic-plugins.md#plugin-manifest-properties).)
> 
> @spadgett, maybe something like this (note all the details in the video are not accurate such as `Enabled`)?
> 
>  Screen.Recording.2024-11-19.at.9.08.21.AM.mov

@rhamilto Yeah, that looks good. We should probably show the current console version in the warning popover and add something to the plugin list table if we don't already have it. We can make it a separate story.

**Comment by spadgett** [comment]
> > 4. I think a table of the extensions used by the plugin would be nice to have in addition to the manifest JSON (but keep the manifest tab for the complete JSON).
> 
> Hmm. Given the variance amongst the extension properties, this one is a bit trickier to tabularize. 🤔

Yeah, I was hesitant about that, too. But even just the list of extension IDs would be helpful. We could show the resource kind for any extension that targets a specific resource and leave the column blank for other extension points. Or maybe not a table, but some other presentation.

**Comment by spadgett** [comment]
> > 3. Syntax highlighting for the manifest JSON would be nice. Maybe we can use Monaco.
> 
> @spadgett, we must've done something wonky with our initial attempt as this appears to be working correctly now.
> 
> I do wonder, is it weird to have two YAML editors?

Yeah, I'm undecided. Maybe we can style it differently or simplify the editor to make it more clear that it's read-only?

All of these suggestions can be follow-up PRs, particularly since this has merged. I'd say the compatible versions tweak is the most important to help admins troubleshoot plugins that aren't enabled.

**Comment by rhamilto** [comment]
> 2\. Should we add a top-level nav items for Console Plugins to the Administration nav section? It would make it more discoverable.

Made this a bug so we can backport:  https://github.com/openshift/console/pull/14521 @spadgett, mind reviewing?

**Pull request merged**